src/main/java/org/breedinginsight/brapps/importer/services/processors/ExperimentProcessor.java

```diff
@@ -0,0 +1,708 @@
+            validateRequiredCell( importRow.getEnvYear(),
+                    "Env Year",
+                    errorMessage, validationErrors, i);
+            validateRequiredCell( importRow.getExpUnitId(),
+                    "Exp Unit ID",
+                    errorMessage, validationErrors, i);
+            validateRequiredCell( importRow.getExpReplicateNo(),
+                    "Exp Replicate #",
+                    errorMessage, validationErrors, i);
+            validateRequiredCell( importRow.getExpBlockNo(),
+                    "Exp Block #",
+                    errorMessage, validationErrors, i);
+        }
+    }
+
+    private void validateRequiredCell(String value, String columnHeader, String errorMessage, ValidationErrors validationErrors, int i) {
+        if ( StringUtils.isBlank( value )) {
+            addRowError(
+                    columnHeader,
+                    errorMessage,
+                    validationErrors, i
+            ) ;
+        }
+    }
+
+    private void addRowError(String field, String errorMessage, ValidationErrors validationErrors, int i) {
+        ValidationError ve = new ValidationError(field, errorMessage, HttpStatus.UNPROCESSABLE_ENTITY);
+        validationErrors.addError(i + 2, ve);  // +2 instead of +1 to account for the column header row.
+    }
+
+    private void addIfNotNull(HashSet<String> set, String setValue) {
+        if( setValue!=null) { set.add(setValue); }
+    }
+
+    private Map<String, ImportPreviewStatistics> getStatisticsMap(List<BrAPIImport> importRows) {
+        // Data for stats.
+        HashSet<String> environmentNameCounter = new HashSet<>(); // set of unique environment names
+        HashSet<String> obsUnitsIDCounter = new HashSet<>(); // set of unique observation unit ID's
+        HashSet<String> gidCounter = new HashSet<>(); // set of unique GID's
+        for (BrAPIImport row : importRows) {
+            ExperimentObservation importRow = (ExperimentObservation) row;
+            // Collect date for stats.
+            addIfNotNull(environmentNameCounter, importRow.getEnv());
+            addIfNotNull(obsUnitsIDCounter, createObservationUnitKey( importRow ));
+            addIfNotNull(gidCounter, importRow.getGid());
+        }
+        ImportPreviewStatistics environmentStats = ImportPreviewStatistics.builder()
+                .newObjectCount(environmentNameCounter.size())
+                .build();
+        ImportPreviewStatistics obdUnitStats = ImportPreviewStatistics.builder()
+                .newObjectCount(obsUnitsIDCounter.size())
+                .build();
+        ImportPreviewStatistics gidStats = ImportPreviewStatistics.builder()
+                .newObjectCount(gidCounter.size())
+                .build();
+
+        return Map.of(
+                "Environments",         environmentStats,
+                "Observation_Units",    obdUnitStats,
+                "GIDs",                 gidStats
+        );
```

Comment:
Should we have the number of new trials in here as well?

Reply:
How about the case that the trial exists but new studies and plots are being added to the trial? It would be useful for the user to know in that case whether a new trial is being created, or if it an existing one was found.